activerecord/lib/active_record/connection_adapters/postgresql/quoting.rb

```diff
@@ -41,9 +41,30 @@ def quote(value) # :nodoc:
           end
         end
 
-        # Quotes strings for use in SQL input.
+        # Quotes strings for use in SQL input. Connection-related exceptions result
+        # in the database reconnecting and retrying the string quoting, up to the
+        # connection's configured +connection_retries+ and +retry_deadline+ limits.
         def quote_string(s) # :nodoc:
+          start_time = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+          retries_available = connection_retries
           valid_raw_connection.escape(s)
+        rescue PG::ConnectionBad => error
+          translated_exception = translate_exception(error, message: error.message, sql: "<escape>", binds: [])
+
+          now = Process.clock_gettime(Process::CLOCK_MONOTONIC)
+          query_time = now - start_time
+          retry_deadline_exceeded = retry_deadline && query_time > retry_deadline
+
+          if !retry_deadline_exceeded && retries_available > 0
+            retries_available -= 1
+
+            if retryable_connection_error?(translated_exception)
+              reconnect!
+              retry
+            end
+          end
```

Comment:
Maybe we should extract this code in a private method in `AbstractAdapter`? I feel like we're coyping it a lot, and it's non-trivial with the deadline logic.

Reply:
Yeah, I was thinking the same thing -- although the code does vary quite a bit between this and the retry logic in `#with_raw_connection`, although there's probably overlap that would make sense to extract.

To keep things simple, I'll wait until https://github.com/rails/rails/pull/46046 is merged so I can clean up the deadline logic to match the changes I just made, and then I'll look into extracting this 😄

Reply:
> the code does vary quite a bit between this and the retry logic in `#with_raw_connection`

Should it?

I came here to claim that [if we want retries] this method should be `with_raw_connection { |c| c.escape(s) }`

Reply:
That makes this much simpler to solve 🤦‍♀️ Thanks Matthew!